File: src/main/java/discretemath/structure/matrix/Matrix.java
package discretemath.structure.matrix;

public interface Matrix<T> {

	int rowsCount();

	int columnsCount();

	T get(int i, int j);

	void set(int i, int j, T value);

	Matrix<T> add(Matrix<T> matrix);

	Matrix<T> multiply(Matrix<T> matrix);

	Matrix<T> transpose();

}

File: src/main/scala/discretemath/structure/matrix/Matrix.scala
package discretemath.structure.matrix;

trait Matrix[T] {

  def rowsCount: Int

  def columnsCount: Int

  def get(i: Int, j: Int): T

  def getRow(i: Int): IndexedSeq[T]

  def set(i: Int, j: Int, value: T): Unit

  def add(matrix: Matrix[T]): Matrix[T]

  def multiply(matrix: Matrix[T]): Matrix[T]

  def transpose: Matrix[T]
}